mars/dataframe/utils.py

```diff
@@ -1464,21 +1455,44 @@ def clean_up_func(op):
             if counted_bytes >= closure_clean_up_bytes_threshold:
                 op.need_clean_up_func = True
                 break
-    # Note: op.func_key is set only when op.need_clean_up_func is True.
+    # Note: op.func_key is set only when func was put into storage.
     if op.need_clean_up_func:
         assert (
             op.logic_key is not None
-        ), "Logic key wasn't calculated before cleaning up func."
-        op.func_key = ctx.storage_put(op.func)
-        op.func = None
+        ), f"Logic key of {op} wasn't calculated before cleaning up func."
+        logger.debug(f"{op} need cleaning up func.")
+        if _is_on_ray(ctx):
+            import ray
+
+            op.func_key = ray.put(op.func)
+            op.func = None
+        else:
+            op.func = cloudpickle.dumps(op.func)
+
+
+def _is_on_ray(ctx):
+    # There are three conditions
+    #   a. mars backend
+    #   b. ray backend(oscar), c. ray backend(dag)
+    # When a. or b. is selected, ctx is an instance of ThreadedServiceContext.
+    #   The main difference between them is whether worker_address matches ray scheme.
+    #   To avoid duplicated checks, here we choose the first worker address.
+    # When c. is selected, ctx is an instance of RayExecutionContext or RayExecutionWorkerContext,
+    #   while get_worker_addresses method isn't currently implemented in RayExecutionWorkerContext.
+    try:
+        worker_addresses = ctx.get_worker_addresses()
+    except AttributeError:  # pragma: no cover
+        assert isinstance(ctx, RayExecutionWorkerContext)
+        return True
+    return isinstance(ctx, RayExecutionContext) or is_ray_address(worker_addresses[0])
 
 
 def restore_func(ctx: Context, op):
```

Comment:
As discussed earlier, add a TODO about removing this function in the future. Construct a wrapped class for functions, override `__getstate__` and `__setstate__` to address these serialization issues.